File: tests/runners/python/tests/06_arrays_and_optionals_test.py
import csv
import json
import tempfile
from pathlib import Path

from .schema import (
    TestCollectionsArrayTest,
    TestCollectionsMixedTestMetadata,
    TestCollectionsOptionalTest,
    TestCollectionsTag,
    load_test_collections_array_tests_from_csv,
    load_test_collections_array_tests_from_json,
    load_test_collections_mixed_tests_from_csv,
    load_test_collections_optional_tests_from_csv,
)


def write_csv(path: Path, rows: list[dict[str, object]]) -> None:
    with path.open("w", encoding="utf-8", newline="") as handle:
        writer = csv.DictWriter(handle, fieldnames=list(rows[0].keys()))
        writer.writeheader()
        for row in rows:
            writer.writerow(row)


def test_from_dict_and_json_loader() -> None:
    with tempfile.TemporaryDirectory() as temp_dir:
        path = Path(temp_dir) / "arrays.json"
        path.write_text(
            json.dumps(
                [
                    {
                        "id": 1,
                        "int_list": [1, -2, 3],
                        "string_list": ["alpha", "beta"],
                        "float_list": [1.5, 2.25],
                        "bool_list": [True, False],
                        "tags": [{"name": "hot", "color": "red"}],
                    }
                ]
            ),
            encoding="utf-8",
        )

        rows = load_test_collections_array_tests_from_json(path)

    assert rows == [
        TestCollectionsArrayTest(
            id=1,
            int_list=[1, -2, 3],
            string_list=["alpha", "beta"],
            float_list=[1.5, 2.25],
            bool_list=[True, False],
            tags=[TestCollectionsTag(name="hot", color="red")],
        )
    ]


def test_csv_loader_parses_lists_optional_and_embeds() -> None:
    with tempfile.TemporaryDirectory() as temp_dir:
        root = Path(temp_dir)
        arrays_csv = root / "arrays.csv"
        optionals_csv = root / "optionals.csv"
        mixed_csv = root / "mixed.csv"

        write_csv(
            arrays_csv,
            [
                {
                    "id": 2,
                    "int_list": "1,-2,3",
                    "string_list": "alpha,beta",
                    "float_list": "1.5,2.25",
                    "bool_list": "yes,no,1,0",
                    "tags": json.dumps([{"name": "cold", "color": "blue"}]),
                }
            ],
        )
        write_csv(
            optionals_csv,
            [
                {
                    "id": 3,
                    "required_name": "optional-row",
                    "opt_int": "",
                    "opt_string": "present",
                    "opt_float": "3.5",
                    "opt_bool": "true",
                    "opt_tag": json.dumps({"name": "maybe", "color": "green"}),
                }
            ],
        )
        write_csv(
            mixed_csv,
            [
                {
                    "id": 4,
                    "opt_tags": json.dumps([{"name": "ui", "color": "white"}]),
                    "meta": json.dumps({"created_by": "alice", "updated_by": None, "version": 7}),
                    "history": json.dumps(
                        [
                            {"created_by": None, "updated_by": "bob", "version": 1},
                            {"created_by": "alice", "updated_by": "carol", "version": 2},
                        ]
                    ),
                }
            ],
        )

        arrays = load_test_collections_array_tests_from_csv(arrays_csv)
        optionals = load_test_collections_optional_tests_from_csv(optionals_csv)
        mixed = load_test_collections_mixed_tests_from_csv(mixed_csv)

    assert arrays[0].bool_list == [True, False, True, False]
    assert arrays[0].tags == [TestCollectionsTag(name="cold", color="blue")]
    assert optionals == [
        TestCollectionsOptionalTest(
            id=3,
            required_name="optional-row",
            opt_int=None,
            opt_string="present",
            opt_float=3.5,
            opt_bool=True,
            opt_tag=TestCollectionsTag(name="maybe", color="green"),
        )
    ]
    assert mixed[0].meta == TestCollectionsMixedTestMetadata(created_by="alice", updated_by=None, version=7)
    assert mixed[0].history[1].updated_by == "carol"


def test_loader_rejects_invalid_values() -> None:
    try:
        TestCollectionsArrayTest.from_dict(
            {
                "id": 1,
                "int_list": ["not-int"],
                "string_list": [],
                "float_list": [],
                "bool_list": [],
                "tags": [],
            }
        )
    except ValueError:
        pass
    else:
        raise AssertionError("invalid integer list item should be rejected")

    try:
        TestCollectionsOptionalTest.from_dict(
            {
                "id": 1,
                "required_name": "bad",
                "opt_int": None,
                "opt_string": None,
                "opt_float": "nan",
                "opt_bool": None,
                "opt_tag": None,
            }
        )
    except ValueError:
        pass
    else:
        raise AssertionError("non-finite optional float should be rejected")


test_from_dict_and_json_loader()
test_csv_loader_parses_lists_optional_and_embeds()
test_loader_rejects_invalid_values()


def test_binary_roundtrip_for_lists_optionals_and_embeds() -> None:
    array_row = TestCollectionsArrayTest(
        id=42,
        int_list=[1, -2, 3],
        string_list=["alpha", "beta"],
        float_list=[1.5, 2.25],
        bool_list=[True, False, True],
        tags=[TestCollectionsTag(name="hot", color="red")],
    )
    loaded_array = TestCollectionsArrayTest.from_binary(array_row.to_binary())
    assert loaded_array.id == 42
    assert loaded_array.int_list == [1, -2, 3]
    assert loaded_array.string_list == ["alpha", "beta"]
    assert loaded_array.bool_list == [True, False, True]
    assert loaded_array.tags == [TestCollectionsTag(name="hot", color="red")]

    optional_row = TestCollectionsOptionalTest(
        id=7,
        required_name="optional",
        opt_int=None,
        opt_string="present",
        opt_float=3.5,
        opt_bool=True,
        opt_tag=TestCollectionsTag(name="maybe", color="green"),
    )
    loaded_optional = TestCollectionsOptionalTest.from_binary(optional_row.to_binary())
    assert loaded_optional == optional_row


test_binary_roundtrip_for_lists_optionals_and_embeds()

File: tests/runners/go/tests/06_arrays_and_optionals_test.go
package polygen

import (
	"bytes"
	"os"
	"path/filepath"
	"testing"
)

func TestGeneratedBinaryIORoundTripArraysAndOptionals(t *testing.T) {
	createdBy := "system"
	updatedBy := "editor"
	row := MixedTest{
		Id: 7,
		OptTags: []Tag{
			{Name: "alpha", Color: "red"},
			{Name: "beta", Color: "blue"},
		},
		Meta: &Metadata{
			CreatedBy: &createdBy,
			Version:   3,
		},
		History: []Metadata{
			{CreatedBy: &createdBy, UpdatedBy: &updatedBy, Version: 1},
			{Version: 2},
		},
	}

	var buf bytes.Buffer
	if err := row.WriteBinary(NewBinaryWriter(&buf)); err != nil {
		t.Fatalf("write mixed binary: %v", err)
	}
	loaded, err := ReadMixedTestBinary(NewBinaryReader(bytes.NewReader(buf.Bytes())))
	if err != nil {
		t.Fatalf("read mixed binary: %v", err)
	}
	if loaded.Id != row.Id || len(loaded.OptTags) != 2 || loaded.OptTags[1].Color != "blue" {
		t.Fatalf("mixed binary tags mismatch: %#v", loaded)
	}
	if loaded.Meta == nil || loaded.Meta.CreatedBy == nil || *loaded.Meta.CreatedBy != "system" || loaded.Meta.Version != 3 {
		t.Fatalf("mixed binary optional metadata mismatch: %#v", loaded.Meta)
	}
	if len(loaded.History) != 2 || loaded.History[0].UpdatedBy == nil || *loaded.History[0].UpdatedBy != "editor" {
		t.Fatalf("mixed binary history mismatch: %#v", loaded.History)
	}
}

func TestGeneratedBinaryTableLoaderRoundTrip(t *testing.T) {
	rows := []*ArrayTest{
		{
			Id:         1,
			IntList:    []int32{1, 2, 3},
			StringList: []string{"a", "b"},
			FloatList:  []float32{1.5, 2.5},
			BoolList:   []bool{true, false, true},
			Tags:       []Tag{{Name: "tag", Color: "green"}},
		},
	}
	path := filepath.Join(t.TempDir(), "array_tests.bin")
	if err := SaveArrayTestsToBinary(path, rows); err != nil {
		t.Fatalf("save array tests: %v", err)
	}
	loaded, err := LoadArrayTestsFromBinary(path)
	if err != nil {
		t.Fatalf("load array tests: %v", err)
	}
	if len(loaded) != 1 || loaded[0].Id != 1 || len(loaded[0].IntList) != 3 || loaded[0].Tags[0].Name != "tag" {
		t.Fatalf("binary table loader returned %#v", loaded)
	}
}

func TestGeneratedCsvLoaderParsesListsAndEmbedJSON(t *testing.T) {
	path := filepath.Join(t.TempDir(), "array_tests.csv")
	content := "id,int_list,string_list,float_list,bool_list,tags\n" +
		"1,\"1, 2,3\",\"alpha,beta\",\"1.25,2.5\",\"yes,no,true\",\"[{\"\"name\"\":\"\"tag-a\"\",\"\"color\"\":\"\"red\"\"},{\"\"name\"\":\"\"tag-b\"\",\"\"color\"\":\"\"blue\"\"}]\"\n"
	if err := os.WriteFile(path, []byte(content), 0644); err != nil {
		t.Fatalf("write csv: %v", err)
	}

	rows, err := LoadArrayTestsFromCsv(path)
	if err != nil {
		t.Fatalf("load array CSV: %v", err)
	}
	if len(rows) != 1 {
		t.Fatalf("expected one row, got %d", len(rows))
	}
	row := rows[0]
	if row.Id != 1 || len(row.IntList) != 3 || row.IntList[2] != 3 {
		t.Fatalf("primitive int list mismatch: %#v", row.IntList)
	}
	if len(row.BoolList) != 3 || !row.BoolList[0] || row.BoolList[1] || !row.BoolList[2] {
		t.Fatalf("bool list mismatch: %#v", row.BoolList)
	}
	if len(row.Tags) != 2 || row.Tags[1].Name != "tag-b" || row.Tags[1].Color != "blue" {
		t.Fatalf("embed JSON list mismatch: %#v", row.Tags)
	}

	badListPath := filepath.Join(t.TempDir(), "bad_array_tests.csv")
	badList := "id,int_list,string_list,float_list,bool_list,tags\n" +
		"1,\"1,nope\",alpha,1.25,true,\"[]\"\n"
	if err := os.WriteFile(badListPath, []byte(badList), 0644); err != nil {
		t.Fatalf("write bad primitive csv: %v", err)
	}
	if _, err := LoadArrayTestsFromCsv(badListPath); err == nil {
		t.Fatalf("expected invalid primitive list item to fail")
	}

	badJSONPath := filepath.Join(t.TempDir(), "bad_tags.csv")
	badJSON := "id,int_list,string_list,float_list,bool_list,tags\n" +
		"1,1,alpha,1.25,true,\"[{bad json}]\"\n"
	if err := os.WriteFile(badJSONPath, []byte(badJSON), 0644); err != nil {
		t.Fatalf("write bad JSON csv: %v", err)
	}
	if _, err := LoadArrayTestsFromCsv(badJSONPath); err == nil {
		t.Fatalf("expected invalid embed JSON list to fail")
	}
}

func TestGeneratedCsvLoaderParsesOptionalAndHistoryEmbedJSON(t *testing.T) {
	path := filepath.Join(t.TempDir(), "mixed_tests.csv")
	content := "id,opt_tags,meta,history\n" +
		"7,\"[{\"\"name\"\":\"\"alpha\"\",\"\"color\"\":\"\"red\"\"}]\",\"{\"\"created_by\"\":\"\"system\"\",\"\"version\"\":3}\",\"[{\"\"updated_by\"\":\"\"editor\"\",\"\"version\"\":1},{\"\"version\"\":2}]\"\n"
	if err := os.WriteFile(path, []byte(content), 0644); err != nil {
		t.Fatalf("write mixed csv: %v", err)
	}

	rows, err := LoadMixedTestsFromCsv(path)
	if err != nil {
		t.Fatalf("load mixed CSV: %v", err)
	}
	if len(rows) != 1 {
		t.Fatalf("expected one mixed row, got %d", len(rows))
	}
	row := rows[0]
	if len(row.OptTags) != 1 || row.OptTags[0].Name != "alpha" {
		t.Fatalf("optional tag list mismatch: %#v", row.OptTags)
	}
	if row.Meta == nil || row.Meta.CreatedBy == nil || *row.Meta.CreatedBy != "system" || row.Meta.Version != 3 {
		t.Fatalf("optional embed JSON mismatch: %#v", row.Meta)
	}
	if len(row.History) != 2 || row.History[0].UpdatedBy == nil || *row.History[0].UpdatedBy != "editor" {
		t.Fatalf("history embed JSON list mismatch: %#v", row.History)
	}
}
